README
```
            __
           /\ \__
 _ __   __ \ \ ,_\   ___   __  _
/\`'__/'__`\\ \ \/  / __`\/\ \/'\
\ \ \/\ \L\.\\ \ \_/\ \L\ \/>  </
 \ \_\ \__/.\_\ \__\ \____//\_/\_\
  \/_/\/__/\/_/\/__/\/___/ \//\/_/
                                   -- by sin and FRIGN
======================================================
```

What is it?
===========

ratox is a client implementation of the rather popular tox protocol[0].
Unlike other clients relying on GUIs as an interface to the user, ratox is
developed with the UNIX-philosophy in mind and allows complete
interaction through named pipes.

There's also a set of scripts[1] developed by various people that build on
top of the FIFO interface.


Getting started
===============

Get the latest version from the git-repository; build and install it.
Run ratox in an empty directory and it will create a set of files and
folders allowing you to control the client.


File structure
==============

A typical filesystem structure is shown below along with some comments
to help explain the semantics of the individual files.

```
.
|-- .ratox.data			# ratox save file
|
|-- 0A734CBA717CEB7883D....	# friend's ID excluding nospam + checksum
|   |-- call_in			# 'arecord -r 48000 -c 1 -f S16_LE > call_in' to initiate a call
|   |-- call_out		# 'aplay -r 48000 -c 1 -f S16_LE - < call_out' to answer a call
|   |-- call_state		# (none, pending, active)
|   |-- file_in			# 'cat foo > file_in' to send a file
|   |-- file_out		# 'cat file_out > bar' to receive a file
|   |-- file_pending		# contains filename if transfer pending, empty otherwise
|   |-- name			# friend's nickname
|   |-- online			# 1 if friend online, 0 otherwise
|   |-- remove			# 'echo 1 > remove' to remove a friend
|   |-- state			# friend's user state; could be any of {none,away,busy}
|   |-- status			# friend's status message
|   |-- text_in			# 'echo yo dude > text_in' to send a text to this friend
|   `-- text_out		# 'tail -f text_out' to dump to stdout any text received
|
|-- id				# 'cat id' to show your own ID, you can give this to your friends
|
|-- name			# changing your nick
|   |-- err			# nickname related errors
|   |-- in			# 'echo my-new-nick > in' to change your name
|   `-- out			# 'cat out' to show your name
|
|-- nospam			# changing your nospam
|   |-- err			# nospam related errors
|   |-- in			# 'echo AABBCCDD > in' to change your nospam
|   `-- out			# 'cat out' to show your nospam
|
|-- request			# send and accept friend requests
|   |-- err			# request related errors
|   |-- in			# 'echo LONGASSID yo dude add me > in' to send a friend request
|   `-- out			# 'echo 1 > out/LONGASSID' to accept the friend request
|
|-- state			# changing your user state
|   |-- err			# user status related errors
|   |-- in			# 'echo away > in' to change your user state; could be any of {none,away,busy}
|   `-- out			# 'cat out' to show your user state
|
`-- status			# changing your status message
    |-- err			# status message related errors
    |-- in			# 'cat I am bored to death > in' to change your status message
    `-- out			# 'cat out' to show your status message
```

Features
========

1 v 1 messaging: Yes
File transfer: Yes
Group chat: No
Audio: Yes
Video: No
DNS discovery: No
Chat logs: Yes
Proxy support: Yes
Offline message: Yes
Offline transfers: Yes
Contact aliases: No
Contact blocking: No
Save file encryption: Yes
Multilingual: No
Multiprofile: Yes
Typing notification: No
Audio notifications: No
Emoticons: No
Spell check: No
Desktop sharing: No
Inline images: No
File resuming: No
Read receipts: No
Message splitting: Yes
Changing nospam: Yes
toxi URI: No

NOTE: Some of these features are not intended to be developed
in ratox itself but rather in external scripts[1] that are built upon
ratox.


Examples
========

SSH over TOX for the practical paranoid
---------------------------------------

On the sender side (the client):
1) cd into the friend's directory (the server)
2) nc -lv 1234 > file_in < file_out

On the receiver side (the server):
1) cd into the friend's directory (the client)
2) cat < file_out | nc localhost 22 > file_in

Now on the client run the following:
ssh -o ProxyCommand="nc %h 1234" user@localhost

Screencasting using ffmpeg and mplayer
--------------------------------------

On the sender side:
ffmpeg -f x11grab -r 10 -s 1366x768 -i :0.0 -vcodec libx264 \
	-pix_fmt yuv420p -preset fast -tune zerolatency -b:v 500k \
	-f flv pipe: > file_in

On the receiver side:
mplayer -cache 1024 file_out

You may have to play about with the cache size.


Portability
===========

Builds and works on *BSD and Linux.  To build on OSX you need
to apply a patch[2].


Contact
=======

You can reach us through the freenode IRC network at #2f30 or
through tech@lists.2f30.org.

[0] https://tox.im/
[1] http://git.2f30.org/ratox-nuggets/
[2] http://ratox.2f30.org/ratox-os_x.patch


debian/changelog
ratox (20160221-1) unstable; urgency=low
  * Updated to new Tox API
  * Disabled AV API temporarily
  

 -- dyngar <cmotc@openmailbox.org>  Sun, 21 Feb 2016 01:32:35 -0500


config.def.h
/* See LICENSE file for copyright and license details. */

/* Connection delay in seconds */
#define CONNECTDELAY 3

/* Ringing delay in seconds */
#define RINGINGDELAY 16

/* Maximum number of simultaneous calls */
#define MAXCALLS 8

static char *savefile        = ".ratox.tox";
static int   encryptsavefile = 0;

static int      ipv6        = 0;
static int      udp         = 1;
static int      proxy       = 0;
static int      proxytype   = 2; /* 1 = HTTP, 2 = SOCKS5 */
static char     proxyaddr[] = "127.0.0.1";
static uint16_t proxyport   = 9050;

static struct node nodes[] = {
	{
		.addr4 = "192.254.75.98",
		.addr6 = "2607:5600:284::2",
		.port  = 33445,
		.idstr = "951C88B7E75C867418ACDB5D273821372BB5BD652740BCDF623A4FA293E75D2F"
	},
	{
		.addr4 = "31.7.57.236",
		.addr6 = "2a02:29b8:1737::fee1:bad:6483:467d",
		.port  = 443,
		.idstr = "2A4B50D1D525DA2E669592A20C327B5FAD6C7E5962DC69296F9FEC77C4436E4E"
	},
	{
		.addr4 = "107.161.17.51",
		.addr6 = "2604:180:1:57:1::",
		.port  = 33445,
		.idstr = "7BE3951B97CA4B9ECDDA768E8C52BA19E9E2690AB584787BF4C90E04DBB75111"
	},
	{
		.addr4 = "144.76.60.215",
		.addr6 = "2a01:4f8:191:64d6::1",
		.port  = 33445,
		.idstr = "04119E835DF3E78BACF0F84235B300546AF8B936F035185E2A8E9E0A67C8924F"
	},
	{
		.addr4 = "23.226.230.47",
		.addr6 = "2604:180:1::3ded:b280",
		.port  = 33445,
		.idstr = "A09162D68618E742FFBCA1C2C70385E6679604B2D80EA6E84AD0996A1AC8A074"
	},
	{
		.addr4 = "178.62.125.224",
		.addr6 = "2a03:b0c0:1:d0::178:6001",
		.port  = 33445,
		.idstr = "10B20C49ACBD968D7C80F2E8438F92EA51F189F4E70CFBBB2C2C8C799E97F03E"
	},
	{
		.addr4 = "37.187.46.132",
		.addr6 = "2001:41d0:0052::0507",
		.port  = 33445,
		.idstr = "5EB67C51D3FF5A9D528D242B669036ED2A30F8A60E674C45E7D43010CB2E1331"
	},
	{
		.addr4 = "178.21.112.187",
		.addr6 = "2a02:2308::216:3eff:fe82:eaef",
		.port  = 33445,
		.idstr = "4B2C19E924972CB9B57732FB172F8A8604DE13EEDA2A6234E348983344B23057"
	},
	{
		.addr4 = "128.199.237.51",
		.addr6 = "2400:6180:0:d0::1ba:7001",
		.port  = 33445,
		.idstr = "67D80D1E00DE2A5294C9CCF9116C72CEC05FB8C2C6DBCEE0B5EB796DA52E2564"
	},
	{
		.addr4 = "195.154.119.113",
		.addr6 = "2001:bc8:3698:101::1",
		.port  = 33445,
		.idstr = "E398A69646B8CEACA9F0B84F553726C1C49270558C57DF5F3C368F05A7D71354"
	},
	{
		.addr4 = "178.62.250.138",
		.addr6 = "2a03:b0c0:2:d0::16:1",
		.port  = 33445,
		.idstr = "788236D34978D1D5BD822F0A5BEBD2C53C64CC31CD3149350EE27D4D9A2F9B6B"
	},
	{
		.addr4 = "178.62.150.106",
		.addr6 = NULL,
		.port  = 33445,
		.idstr = "58D2DE4B169502669941E50780C1630FAA48A0B7026D6F4066C320D47AC6401E"
	},
	{
		.addr4 = "192.210.149.121",
		.addr6 = NULL,
		.port  = 33445,
		.idstr = "F404ABAA1C99A9D37D61AB54898F56793E1DEF8BD46B1038B9D822E8460FAB67"
	},
	{
		.addr4 = "54.199.139.199",
		.addr6 = NULL,
		.port  = 33445,
		.idstr = "7F9C31FE850E97CEFD4C4591DF93FC757C7C12549DDD55F8EEAECC34FE76C029"
	},
	{
		.addr4 = "63.165.243.15",
		.addr6 = NULL,
		.port  = 443,
		.idstr = "8CD087E31C67568103E8C2A28653337E90E6B8EDA0D765D57C6B5172B4F1F04C"
	},
	{
		.addr4 = "76.191.23.96",
		.addr6 = NULL,
		.port  = 33445,
		.idstr = "93574A3FAB7D612FEA29FD8D67D3DD10DFD07A075A5D62E8AF3DD9F5D0932E11"
	},
	{
		.addr4 = "46.38.239.179",
		.addr6 = NULL,
		.port  = 33445,
		.idstr = "F5A1A38EFB6BD3C2C8AF8B10D85F0F89E931704D349F1D0720C3C4059AF2440A"
	},
	{
		.addr4 = "144.76.93.230",
		.addr6 = NULL,
		.port  = 33445,
		.idstr = "2C308B4518862740AD9A121598BCA7713AFB25858B747313A4D073E2F6AC506C"
	},
	{
		.addr4 = "78.225.128.39",
		.addr6 = NULL,
		.port  = 33445,
		.idstr = "7A2306BFBA665E5480AE59B31E116BE9C04DCEFE04D9FE25082316FA34B4DA0C"
	},
	{
		.addr4 = "195.154.109.148",
		.addr6 = NULL,
		.port  = 33445,
		.idstr = "391C96CB67AE893D4782B8E4495EB9D89CF1031F48460C06075AA8CE76D50A21"
	},
	{
		.addr4 = "192.3.173.88",
		.addr6 = NULL,
		.port  = 33445,
		.idstr = "3E1FFDEB667BFF549F619EC6737834762124F50A89C8D0DBF1DDF64A2DD6CD1B"
	},
	{
		.addr4 = "130.133.110.14",
		.addr6 = "2001:6f8:1c3c:babe::14:1",
		.port  = 33445,
		.idstr = "461FA3776EF0FA655F1A05477DF1B3B614F7D6B124F7DB1DD4FE3C08B03B640F"
	},
};


config.h
/* See LICENSE file for copyright and license details. */

/* Connection delay in seconds */
#define CONNECTDELAY 3

/* Ringing delay in seconds */
#define RINGINGDELAY 16

/* Maximum number of simultaneous calls */
#define MAXCALLS 8

static char *savefile        = ".ratox.tox";
static int   encryptsavefile = 0;

static int      ipv6        = 0;
static int      tcp         = 0;
static int      proxy       = 0;
static int      proxytype   = 2; /* 1 = HTTP, 2 = SOCKS5 */
static char     proxyaddr[] = "127.0.0.1";
static uint16_t proxyport   = 9050;

static struct node nodes[] = {
	{
		.addr4 = "192.254.75.98",
		.addr6 = "2607:5600:284::2",
		.port  = 33445,
		.idstr = "951C88B7E75C867418ACDB5D273821372BB5BD652740BCDF623A4FA293E75D2F"
	},
	{
		.addr4 = "31.7.57.236",
		.addr6 = "2a02:29b8:1737::fee1:bad:6483:467d",
		.port  = 443,
		.idstr = "2A4B50D1D525DA2E669592A20C327B5FAD6C7E5962DC69296F9FEC77C4436E4E"
	},
	{
		.addr4 = "107.161.17.51",
		.addr6 = "2604:180:1:57:1::",
		.port  = 33445,
		.idstr = "7BE3951B97CA4B9ECDDA768E8C52BA19E9E2690AB584787BF4C90E04DBB75111"
	},
	{
		.addr4 = "144.76.60.215",
		.addr6 = "2a01:4f8:191:64d6::1",
		.port  = 33445,
		.idstr = "04119E835DF3E78BACF0F84235B300546AF8B936F035185E2A8E9E0A67C8924F"
	},
	{
		.addr4 = "23.226.230.47",
		.addr6 = "2604:180:1::3ded:b280",
		.port  = 33445,
		.idstr = "A09162D68618E742FFBCA1C2C70385E6679604B2D80EA6E84AD0996A1AC8A074"
	},
	{
		.addr4 = "178.62.125.224",
		.addr6 = "2a03:b0c0:1:d0::178:6001",
		.port  = 33445,
		.idstr = "10B20C49ACBD968D7C80F2E8438F92EA51F189F4E70CFBBB2C2C8C799E97F03E"
	},
	{
		.addr4 = "37.187.46.132",
		.addr6 = "2001:41d0:0052::0507",
		.port  = 33445,
		.idstr = "5EB67C51D3FF5A9D528D242B669036ED2A30F8A60E674C45E7D43010CB2E1331"
	},
	{
		.addr4 = "178.21.112.187",
		.addr6 = "2a02:2308::216:3eff:fe82:eaef",
		.port  = 33445,
		.idstr = "4B2C19E924972CB9B57732FB172F8A8604DE13EEDA2A6234E348983344B23057"
	},
	{
		.addr4 = "128.199.237.51",
		.addr6 = "2400:6180:0:d0::1ba:7001",
		.port  = 33445,
		.idstr = "67D80D1E00DE2A5294C9CCF9116C72CEC05FB8C2C6DBCEE0B5EB796DA52E2564"
	},
	{
		.addr4 = "195.154.119.113",
		.addr6 = "2001:bc8:3698:101::1",
		.port  = 33445,
		.idstr = "E398A69646B8CEACA9F0B84F553726C1C49270558C57DF5F3C368F05A7D71354"
	},
	{
		.addr4 = "178.62.250.138",
		.addr6 = "2a03:b0c0:2:d0::16:1",
		.port  = 33445,
		.idstr = "788236D34978D1D5BD822F0A5BEBD2C53C64CC31CD3149350EE27D4D9A2F9B6B"
	},
	{
		.addr4 = "178.62.150.106",
		.addr6 = NULL,
		.port  = 33445,
		.idstr = "58D2DE4B169502669941E50780C1630FAA48A0B7026D6F4066C320D47AC6401E"
	},
	{
		.addr4 = "192.210.149.121",
		.addr6 = NULL,
		.port  = 33445,
		.idstr = "F404ABAA1C99A9D37D61AB54898F56793E1DEF8BD46B1038B9D822E8460FAB67"
	},
	{
		.addr4 = "54.199.139.199",
		.addr6 = NULL,
		.port  = 33445,
		.idstr = "7F9C31FE850E97CEFD4C4591DF93FC757C7C12549DDD55F8EEAECC34FE76C029"
	},
	{
		.addr4 = "63.165.243.15",
		.addr6 = NULL,
		.port  = 443,
		.idstr = "8CD087E31C67568103E8C2A28653337E90E6B8EDA0D765D57C6B5172B4F1F04C"
	},
	{
		.addr4 = "76.191.23.96",
		.addr6 = NULL,
		.port  = 33445,
		.idstr = "93574A3FAB7D612FEA29FD8D67D3DD10DFD07A075A5D62E8AF3DD9F5D0932E11"
	},
	{
		.addr4 = "46.38.239.179",
		.addr6 = NULL,
		.port  = 33445,
		.idstr = "F5A1A38EFB6BD3C2C8AF8B10D85F0F89E931704D349F1D0720C3C4059AF2440A"
	},
	{
		.addr4 = "144.76.93.230",
		.addr6 = NULL,
		.port  = 33445,
		.idstr = "2C308B4518862740AD9A121598BCA7713AFB25858B747313A4D073E2F6AC506C"
	},
	{
		.addr4 = "78.225.128.39",
		.addr6 = NULL,
		.port  = 33445,
		.idstr = "7A2306BFBA665E5480AE59B31E116BE9C04DCEFE04D9FE25082316FA34B4DA0C"
	},
	{
		.addr4 = "195.154.109.148",
		.addr6 = NULL,
		.port  = 33445,
		.idstr = "391C96CB67AE893D4782B8E4495EB9D89CF1031F48460C06075AA8CE76D50A21"
	},
	{
		.addr4 = "192.3.173.88",
		.addr6 = NULL,
		.port  = 33445,
		.idstr = "3E1FFDEB667BFF549F619EC6737834762124F50A89C8D0DBF1DDF64A2DD6CD1B"
	},
	{
		.addr4 = "130.133.110.14",
		.addr6 = "2001:6f8:1c3c:babe::14:1",
		.port  = 33445,
		.idstr = "461FA3776EF0FA655F1A05477DF1B3B614F7D6B124F7DB1DD4FE3C08B03B640F"
	},
};
